external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala

```diff
@@ -18,60 +18,206 @@
+
+  override val brokerProps = Map("auto.create.topics.enable" -> "false")
+
+  override def afterAll(): Unit = {
+    if (testUtils != null) {
+      testUtils.teardown()
+      testUtils = null
+    }
+    super.afterAll()
+  }
+
+  override def sparkConf: SparkConf = {
+    val conf = super.sparkConf
+    conf.set("spark.kafka.producer.cache.timeout", "2ms")
+  }
+
+  test("concurrent use of CachedKafkaProducer") {
+    val topic = "topic" + Random.nextInt()
+    val data = (1 to 10).map(_.toString)
+    testUtils.createTopic(topic, 1)
+    val kafkaParams: Map[String, Object] = Map("bootstrap.servers" -> testUtils.brokerAddress,
+      "key.serializer" -> classOf[ByteArraySerializer].getName,
+      "value.serializer" -> classOf[ByteArraySerializer].getName)
 
-    CachedKafkaProducer.close(kafkaParams)
-    val map2 = CachedKafkaProducer.invokePrivate(cacheMap())
-    assert(map2.size == 1)
     import scala.collection.JavaConverters._
-    val (seq: Seq[(String, Object)], _producer: KP) = map2.asScala.toArray.apply(0)
-    assert(_producer == producer)
+
+    val numThreads: Int = 100
+    val numConcurrentProducers: Int = 1000
+
+    val kafkaParamsUniqueMap = mutable.HashMap.empty[Int, ju.Map[String, Object]]
+    ( 1 to numConcurrentProducers).map {
+      i => kafkaParamsUniqueMap.put(i, kafkaParams.updated("retries", s"$i").asJava)
+    }
+    val toBeReleasedQueue = new ConcurrentLinkedQueue[CachedKafkaProducer]()
+
+    def acquire(i: Int): Unit = {
+      val producer = CachedKafkaProducer.acquire(kafkaParamsUniqueMap(i))
+      producer.kafkaProducer // materialize producer for the first time.
+      assert(!producer.isClosed, "Acquired producer cannot be closed.")
+      toBeReleasedQueue.add(producer)
+    }
+
+    def release(producer: CachedKafkaProducer): Unit = {
+      if (producer != null) {
+        CachedKafkaProducer.release(producer, true)
+        if (producer.getInUseCount > 0) {
+          assert(!producer.isClosed, "Should not close an inuse producer.")
+        }
+      }
+    }
+    val threadPool = Executors.newFixedThreadPool(numThreads)
+    try {
+      val futuresAcquire = (1 to 10 * numConcurrentProducers).map { i =>
+        threadPool.submit(new Runnable {
+          override def run(): Unit = {
+            acquire(i % numConcurrentProducers + 1)
+          }
+        })
+      }
+      val futuresRelease = (1 to 10 * numConcurrentProducers).map { i =>
+        val cachedKafkaProducer = toBeReleasedQueue.poll()
+        // 2x release should not corrupt the state of cache.
+        (1 to 2).map { j =>
+          threadPool.submit(new Runnable {
+            override def run(): Unit = {
+              release(cachedKafkaProducer)
+            }
+          })
+        }
+      }
+      futuresAcquire.foreach(_.get(1, TimeUnit.MINUTES))
+      futuresRelease.flatten.foreach(_.get(1, TimeUnit.MINUTES))
+    } finally {
+      threadPool.shutdown()
```

Comment:
`CachedKafkaProducer.clear()` would be good somewhere because proper clean happens only on happy path.